File: test/ts/ssh-test/metricsTests.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as assert from 'assert';
import { suite, test, slow, timeout, pending, params } from '@testdeck/mocha';

import {
	KeyPair,
	SshAlgorithms,
	SshClientSession,
	SshServerSession,
	SshChannel,
	SshDisconnectReason,
	PromiseCompletionSource,
	SshSessionConfiguration,
	SshProtocolExtensionNames,
	SessionContour,
} from '@microsoft/dev-tunnels-ssh';
import {
	createSessionPair,
	connectSessionPair,
	authenticateClient,
	authenticateServer,
} from './sessionPair';
import { SessionMetrics } from 'src/ts/ssh/metrics/sessionMetrics';
import { withTimeout } from './promiseUtils';

@suite
@slow(3000)
@timeout(10000)
export class MetricsTests {
	private static serverKey: KeyPair;

	@slow(10000)
	@timeout(20000)
	public static async before() {
		MetricsTests.serverKey = await SshAlgorithms.publicKey.ecdsaSha2Nistp384!.generateKeyPair();
	}

	public static async createSessions(): Promise<[SshClientSession, SshServerSession]> {
		const config = new SshSessionConfiguration();
		config.protocolExtensions.push(SshProtocolExtensionNames.sessionReconnect);
		config.protocolExtensions.push(SshProtocolExtensionNames.sessionLatency);

		const serverSession = new SshServerSession(config, []);
		const clientSession = new SshClientSession(config);

		authenticateClient(clientSession, serverSession);
		authenticateServer(clientSession, serverSession, MetricsTests.serverKey);

		await connectSessionPair(clientSession, serverSession);

		const authenticated = await clientSession.authenticate({ username: 'test' });
		assert(authenticated);

		return [clientSession, serverSession];
	}

	@test
	public async measureChannelBytes() {
		const [clientSession, serverSession] = await MetricsTests.createSessions();
		const serverChannelTask = serverSession.acceptChannel();
		const clientChannel = await clientSession.openChannel();
		const serverChannel = await serverChannelTask;

		await MetricsTests.sendDataFromClientToServerChannel(
			[Buffer.from('A', 'utf8'), Buffer.from('abc', 'utf8')],
			clientChannel,
			serverChannel,
		);

		assert.equal(clientChannel.metrics.bytesSent, 4);
		assert.equal(clientChannel.metrics.bytesReceived, 0);
		assert.equal(serverChannel.metrics.bytesSent, 0);
		assert.equal(serverChannel.metrics.bytesReceived, 4);
	}

	@test
	public async measureSessionBytes() {
		const [clientSession, serverSession] = await MetricsTests.createSessions();
		const serverChannelTask = serverSession.acceptChannel();
		const clientChannel = await clientSession.openChannel();
		const serverChannel = await serverChannelTask;

		const initialClientBytesSent = clientSession.metrics.bytesSent;
		const initialClientBytesReceived = clientSession.metrics.bytesReceived;
		const initialServerBytesSent = serverSession.metrics.bytesSent;
		const initialServerBytesReceived = serverSession.metrics.bytesReceived;

		assert.notEqual(initialClientBytesSent, 0);
		assert.notEqual(initialClientBytesReceived, 0);
		assert.notEqual(initialServerBytesSent, 0);
		assert.notEqual(initialServerBytesReceived, 0);

		await MetricsTests.sendDataFromClientToServerChannel(
			[Buffer.from('A', 'utf8'), Buffer.from('abc', 'utf8')],
			clientChannel,
			serverChannel,
		);

		assert(clientSession.metrics.bytesSent > initialClientBytesSent);
		assert.equal(clientSession.metrics.bytesReceived, initialClientBytesReceived);
		assert.equal(serverSession.metrics.bytesSent, initialServerBytesSent);
		assert(serverSession.metrics.bytesReceived > initialServerBytesReceived);
	}

	@test
	public async measureSessionMessages() {
		const [clientSession, serverSession] = await MetricsTests.createSessions();
		const serverChannelTask = serverSession.acceptChannel();
		const clientChannel = await clientSession.openChannel();
		const serverChannel = await serverChannelTask;

		const initialClientMessagesSent = clientSession.metrics.messagesSent;
		const initialClientMessagesReceived = clientSession.metrics.messagesReceived;
		const initialServerMessagesSent = serverSession.metrics.messagesSent;
		const initialServerMessagesReceived = serverSession.metrics.messagesReceived;

		assert.notEqual(initialClientMessagesSent, 0);
		assert.notEqual(initialClientMessagesReceived, 0);
		assert.notEqual(initialServerMessagesSent, 0);
		assert.notEqual(initialServerMessagesReceived, 0);

		await MetricsTests.sendDataFromClientToServerChannel(
			[Buffer.from('A', 'utf8'), Buffer.from('abc', 'utf8')],
			clientChannel,
			serverChannel,
		);

		assert(clientSession.metrics.messagesSent > initialClientMessagesSent);
		assert.equal(clientSession.metrics.messagesReceived, initialClientMessagesReceived);
		assert.equal(serverSession.metrics.messagesSent, initialServerMessagesSent);
		assert(serverSession.metrics.messagesReceived > initialServerMessagesReceived);
	}

	@test
	public async measureSessionLatency() {
		const [clientSession, serverSession] = await MetricsTests.createSessions();
		const serverChannelTask = serverSession.acceptChannel();
		const clientChannel = await clientSession.openChannel();
		const serverChannel = await serverChannelTask;

		for (let i = 0; i < 100; i++) {
			await MetricsTests.sendDataFromClientToServerChannel(
				[Buffer.from('abc'.repeat(1000), 'utf8')],
				clientChannel,
				serverChannel,
			);
		}

		function validateLatency(metrics: SessionMetrics): void {
			assert.notEqual(metrics.latencyMinMs, 0);
			assert.notEqual(metrics.latencyAverageMs, 0);
			assert.notEqual(metrics.latencyMaxMs, 0);
			assert.notEqual(metrics.latencyCurrentMs, 0);
			assert(metrics.latencyMinMs <= metrics.latencyAverageMs);
			assert(metrics.latencyAverageMs <= metrics.latencyMaxMs);
		}

		validateLatency(clientSession.metrics);
		validateLatency(serverSession.metrics);
	}

	@test
	public async closedSessionHasNoLatency() {
		const [clientSession, serverSession] = await MetricsTests.createSessions();
		const serverChannelTask = serverSession.acceptChannel();
		await clientSession.openChannel();
		await serverChannelTask;

		await clientSession.close(SshDisconnectReason.byApplication);
		await serverSession.close(SshDisconnectReason.byApplication);

		assert.equal(clientSession.metrics.latencyCurrentMs, 0);
		assert.equal(serverSession.metrics.latencyCurrentMs, 0);
	}

	@test
	@slow(7000)
	public async recordSessionContour() {
		const [clientSession, serverSession] = await MetricsTests.createSessions();

		const clientContour = new SessionContour(256);
		const serverContour = new SessionContour(256);

		const clientContourPromise = clientContour.collectMetrics(clientSession.metrics);
		const serverContourPromise = serverContour.collectMetrics(serverSession.metrics);

		const validateContour = (contour: SessionContour) => {
			// Normally the interval should be 1 second, but tests tests need to work
			// on very slow build machines where the interval could grow larger.
			assert([1000, 2000, 4000].indexOf(contour.interval) >= 0);

			assert.notEqual(contour.intervalCount, 0);
			const sum = (array: readonly number[]) => array.reduce((a, b) => a + b, 0);
			assert.notEqual(sum(contour.latencyMinMs), 0);
			assert.notEqual(sum(contour.latencyMaxMs), 0);
			assert.notEqual(sum(contour.latencyAverageMs), 0);
			assert.notEqual(sum(contour.bytesSent), 0);
			assert.notEqual(sum(contour.bytesReceived), 0);

			for (let i = 0; i < contour.intervalCount; i++) {
				assert(contour.latencyMinMs[i] <= contour.latencyAverageMs[i]);
				assert(contour.latencyAverageMs[i] <= contour.latencyMaxMs[i]);
			}
		};

		const serverChannelTask = serverSession.acceptChannel();
		const clientChannel = await clientSession.openChannel();
		const serverChannel = await serverChannelTask;

		await new Promise((c) => setTimeout(c, 1000));
		const data = [Buffer.from('A', 'utf8')];
		await MetricsTests.sendDataFromClientToServerChannel(data, clientChannel, serverChannel);
		await MetricsTests.sendDataFromClientToServerChannel(data, serverChannel, clientChannel);
		await new Promise((c) => setTimeout(c, 1000));
		await MetricsTests.sendDataFromClientToServerChannel(data, clientChannel, serverChannel);
		await MetricsTests.sendDataFromClientToServerChannel(data, serverChannel, clientChannel);
		await new Promise((c) => setTimeout(c, 1000));
		await MetricsTests.sendDataFromClientToServerChannel(data, clientChannel, serverChannel);
		await MetricsTests.sendDataFromClientToServerChannel(data, serverChannel, clientChannel);

		await MetricsTests.waitForContourUpdate(clientContour, clientContourPromise);
		await MetricsTests.waitForContourUpdate(serverContour, serverContourPromise);

		validateContour(clientContour);
		validateContour(serverContour);
	}

	@test
	public async expandContourIntervals() {
		const session = new SshClientSession(new SshSessionConfiguration());
		const metrics = session.metrics;

		const sessionContour = new SessionContour(4);
		let updatePromise = sessionContour.collectMetrics(metrics);
		MetricsTests.addMessageReceived(sessionContour, 2000, 2);
		MetricsTests.updateLatency(sessionContour, 3000, 16);
		MetricsTests.updateLatency(sessionContour, 3500, 32);
		MetricsTests.addMessageSent(sessionContour, 3800, 1);
		MetricsTests.addMessageReceived(sessionContour, 3900, 3);
		await MetricsTests.waitForContourUpdate(sessionContour, updatePromise);
		assert.equal(sessionContour.interval, 1000);
		assert.deepEqual(sessionContour.latencyMinMs, [0, 0, 0, 16]);
		assert.deepEqual(sessionContour.latencyMaxMs, [0, 0, 0, 32]);
		assert.deepEqual(sessionContour.latencyAverageMs, [0, 0, 0, 24]);
		assert.deepEqual(sessionContour.bytesSent, [0, 0, 0, 1]);
		assert.deepEqual(sessionContour.bytesReceived, [0, 0, 2, 3]);
		updatePromise = sessionContour.collectMetrics(metrics);
		MetricsTests.addMessageSent(sessionContour, 4000, 1);
		MetricsTests.updateLatency(sessionContour, 4500, 32);
		MetricsTests.updateLatency(sessionContour, 4600, 16);
		await MetricsTests.waitForContourUpdate(sessionContour, updatePromise);
		assert.equal(sessionContour.interval, 2000);
		assert.deepEqual(sessionContour.latencyMinMs, [0, 16, 16]);
		assert.deepEqual(sessionContour.latencyMaxMs, [0, 32, 32]);
		assert.deepEqual(sessionContour.latencyAverageMs, [0, 24, 24]);
		assert.deepEqual(sessionContour.bytesSent, [0, 1, 1]);
		assert.deepEqual(sessionContour.bytesReceived, [0, 5, 0]);
		updatePromise = sessionContour.collectMetrics(metrics);
		MetricsTests.addMessageSent(sessionContour, 8000, 1);
		MetricsTests.updateLatency(sessionContour, 8100, 32);
		MetricsTests.addMessageSent(sessionContour, 12000, 2);
		MetricsTests.updateLatency(sessionContour, 12500, 64);
		await MetricsTests.waitForContourUpdate(sessionContour, updatePromise);
		assert.equal(sessionContour.interval, 4000);
		assert.deepEqual(sessionContour.latencyMinMs, [16, 16, 32, 64]);
		assert.deepEqual(sessionContour.latencyMaxMs, [32, 32, 32, 64]);
		assert.deepEqual(sessionContour.latencyAverageMs, [24, 24, 32, 64]);
		assert.deepEqual(sessionContour.bytesSent, [1, 1, 1, 2]);
		assert.deepEqual(sessionContour.bytesReceived, [5, 0, 0, 0]);
		updatePromise = sessionContour.collectMetrics(metrics);
		MetricsTests.addMessageSent(sessionContour, 16000, 10);
		await MetricsTests.waitForContourUpdate(sessionContour, updatePromise);
		assert.equal(sessionContour.interval, 8000);
		assert.deepEqual(sessionContour.latencyMinMs, [16, 32, 0]);
		assert.deepEqual(sessionContour.latencyMaxMs, [32, 64, 0]);
		assert.deepEqual(sessionContour.latencyAverageMs, [24, 48, 0]);
		assert.deepEqual(sessionContour.bytesSent, [2, 3, 10]);
		assert.deepEqual(sessionContour.bytesReceived, [5, 0, 0]);
	}

	@test
	public async exportImportContour() {
		const session = new SshClientSession(new SshSessionConfiguration());
		const metrics = session.metrics;

		const sessionContour = new SessionContour(4);
		let updatePromise = sessionContour.collectMetrics(metrics);
		MetricsTests.addMessageReceived(sessionContour, 0, 2000);
		MetricsTests.updateLatency(sessionContour, 2000, 16);
		MetricsTests.updateLatency(sessionContour, 3000, 32);
		MetricsTests.addMessageSent(sessionContour, 3600, 1000);
		MetricsTests.addMessageReceived(sessionContour, 3800, 3000);
		MetricsTests.addMessageSent(sessionContour, 4800, 1);
		MetricsTests.updateLatency(sessionContour, 5000, 32);
		MetricsTests.updateLatency(sessionContour, 5200, 16);
		await MetricsTests.waitForContourUpdate(sessionContour, updatePromise);
		assert.equal(sessionContour.interval, 2000);

		const result = sessionContour.export();
		const resultBytes = Array.from(Buffer.from(result, 'base64'));

		assert.deepEqual(resultBytes, [
			1, // version
			5, // metric count
			1, // timeScale
			0, // \
			0, //  \
			0, //   } value scales
			2, //  /
			4, // /
			1, // \
			2, //  \
			3, //   } metric IDs
			11, //  /
			12, // /
			0, // \
			0, //  \
			0, //   } interval 0
			0, //  /
			125, // /
			16, // \
			32, //  \
			24, //   } interval 1
			250, //  /
			188, // /
			16, // \
			32, //  \
			24, //   } interval 2
			0, //  /
			0, // /
		]);

		const sessionContour2 = SessionContour.import(result);
		assert.equal(sessionContour2.intervalCount, 3);
		assert.equal(2000, sessionContour2.interval);

		const result2 = sessionContour2.export();
		const result2Bytes = Array.from(Buffer.from(result2, 'base64'));
		assert.deepEqual(result2Bytes, resultBytes);
	}

	private static addMessageSent(sessionContour: SessionContour, time: number, size: number): void {
		(<any>sessionContour).onMessageSent({ time, size });
	}

	private static addMessageReceived(
		sessionContour: SessionContour,
		time: number,
		size: number,
	): void {
		(<any>sessionContour).onMessageReceived({ time, size });
	}

	private static updateLatency(
		sessionContour: SessionContour,
		time: number,
		latency: number,
	): void {
		(<any>sessionContour).onLatencyUpdated({ time, latency });
	}

	private static async waitForContourUpdate(
		sessionContour: SessionContour,
		p: Promise<void>,
	): Promise<void> {
		(<any>sessionContour).onSessionClosed();
		await withTimeout(p, 5000);
	}

	private static async sendDataFromClientToServerChannel(
		data: Buffer[],
		clientChannel: SshChannel,
		serverChannel: SshChannel,
	): Promise<void> {
		let receivingBuffer: Buffer | null = null;
		let receivedCompletion: PromiseCompletionSource<Buffer> | null = null;
		let expectedDataLength = 0;

		const dataReceivedRegistration = serverChannel.onDataReceived(async (received) => {
			receivingBuffer = Buffer.concat([receivingBuffer!, received]);
			if (receivingBuffer.length >= expectedDataLength) {
				receivedCompletion!.resolve(receivingBuffer);
			}

			serverChannel.adjustWindow(received.length);
		});

		for (let i = 0; i < data.length; i++) {
			receivingBuffer = Buffer.alloc(0);
			receivedCompletion = new PromiseCompletionSource<Buffer>();
			expectedDataLength = data[i].length;

			await clientChannel.send(data[i]);

			const receivedData = await receivedCompletion.promise;
			assert.equal(Buffer.compare(data[i], receivedData), 0);
		}

		dataReceivedRegistration.dispose();
	}
}

File: test/cs/Ssh.Test/MetricsTests.cs
using Microsoft.DevTunnels.Ssh.Metrics;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace Microsoft.DevTunnels.Ssh.Test;

public class MetricsTests : IDisposable
{
	private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(10);

	private static readonly BindingFlags PrivateInstanceBinding =
		BindingFlags.NonPublic | BindingFlags.Instance;
	private static readonly MethodInfo SessionContourOnMessageSentMethod =
		typeof(SessionContour).GetMethod("OnMessageSent", PrivateInstanceBinding);
	private static readonly MethodInfo SessionContourOnMessageReceivedMethod =
		typeof(SessionContour).GetMethod("OnMessageReceived", PrivateInstanceBinding);
	private static readonly MethodInfo SessionContourOnLatencyUpdatedMethod =
		typeof(SessionContour).GetMethod("OnLatencyUpdated", PrivateInstanceBinding);
	private static readonly MethodInfo SessionContourSessionClosedMethod =
		typeof(SessionContour).GetMethod("OnSessionClosed", PrivateInstanceBinding);

	private SessionPair sessionPair;
	private SshServerSession serverSession;
	private SshClientSession clientSession;
	private SessionContour sessionContour;

	public MetricsTests()
	{
		InitializeSessionPair();
	}

	private void InitializeSessionPair()
	{
		var serverConfig = SshSessionConfiguration.DefaultWithReconnect;
		var clientConfig = SshSessionConfiguration.DefaultWithReconnect;

		var reconnectableSessions = new List<SshServerSession>();
		this.sessionPair = new SessionPair(serverConfig, clientConfig, reconnectableSessions);
		this.serverSession = this.sessionPair.ServerSession;
		this.clientSession = this.sessionPair.ClientSession;
		this.sessionContour = new SessionContour(4);
	}

	public void Dispose()
	{
		this.sessionPair.Dispose();
	}

	[Fact]
	public async Task MeasureChannelBytes()
	{
		var channels = await OpenClientChannelAsync();

		byte[][] data = new[] { new byte[] { 1 }, new byte[] { 1, 2, 3 } };
		await SendDataBetweenChannelsAsync(data, channels.Client, channels.Server);

		Assert.Equal(4, channels.Client.Metrics.BytesSent);
		Assert.Equal(0, channels.Client.Metrics.BytesReceived);
		Assert.Equal(0, channels.Server.Metrics.BytesSent);
		Assert.Equal(4, channels.Server.Metrics.BytesReceived);
	}

	[Fact]
	public async Task MeasureSessionBytes()
	{
		var channels = await OpenClientChannelAsync();

		var initialClientBytesSent = this.clientSession.Metrics.BytesSent;
		var initialClientBytesReceived = this.clientSession.Metrics.BytesReceived;
		var initialServerBytesSent = this.serverSession.Metrics.BytesSent;
		var initialServerBytesReceived = this.serverSession.Metrics.BytesReceived;
		Assert.NotEqual(0, initialClientBytesSent);
		Assert.NotEqual(0, initialClientBytesReceived);
		Assert.NotEqual(0, initialServerBytesSent);
		Assert.NotEqual(0, initialServerBytesReceived);

		byte[][] data = new[] { new byte[] { 1 }, new byte[] { 1, 2, 3 } };
		await SendDataBetweenChannelsAsync(data, channels.Client, channels.Server);

		Assert.True(this.clientSession.Metrics.BytesSent > initialClientBytesSent);
		Assert.Equal(initialClientBytesReceived, this.clientSession.Metrics.BytesReceived);
		Assert.Equal(initialServerBytesSent, this.serverSession.Metrics.BytesSent);
		Assert.True(this.serverSession.Metrics.BytesReceived > initialServerBytesReceived);
	}

	[Fact]
	public async Task MeasureSessionMessages()
	{
		var channels = await OpenClientChannelAsync();

		var initialClientMessagesSent = this.clientSession.Metrics.MessagesSent;
		var initialClientMessagesReceived = this.clientSession.Metrics.MessagesReceived;
		var initialServerMessagesSent = this.serverSession.Metrics.MessagesSent;
		var initialServerMessagesReceived = this.serverSession.Metrics.MessagesReceived;
		Assert.NotEqual(0, initialClientMessagesSent);
		Assert.NotEqual(0, initialClientMessagesReceived);
		Assert.NotEqual(0, initialServerMessagesSent);
		Assert.NotEqual(0, initialServerMessagesReceived);

		byte[][] data = new[] { new byte[] { 1 }, new byte[] { 1, 2, 3 } };
		await SendDataBetweenChannelsAsync(data, channels.Client, channels.Server);

		Assert.True(this.clientSession.Metrics.MessagesSent > initialClientMessagesSent);
		Assert.Equal(initialClientMessagesReceived, this.clientSession.Metrics.MessagesReceived);
		Assert.Equal(initialServerMessagesSent, this.serverSession.Metrics.MessagesSent);
		Assert.True(this.serverSession.Metrics.MessagesReceived > initialServerMessagesReceived);
	}

	[Fact]
	public async Task MeasureSessionLatency()
	{
		var channels = await OpenClientChannelAsync();
		byte[][] data = new[] { new byte[] { 1 } };
		await SendDataBetweenChannelsAsync(data, channels.Client, channels.Server);
		await SendDataBetweenChannelsAsync(data, channels.Server, channels.Client);

		void ValidateLatency(SessionMetrics metrics)
		{
			Assert.NotEqual(0, metrics.LatencyMaxMs);
			Assert.NotEqual(0, metrics.LatencyAverageMs);
			Assert.NotEqual(0, metrics.LatencyMinMs);
			Assert.NotEqual(0, metrics.LatencyCurrentMs);
			Assert.True(metrics.LatencyMinMs <= metrics.LatencyAverageMs);
			Assert.True(metrics.LatencyAverageMs <= metrics.LatencyMaxMs);
		}

		ValidateLatency(this.clientSession.Metrics);
		ValidateLatency(this.serverSession.Metrics);
	}

	[Fact]
	public async Task ClosedSessionHasNoLatency()
	{
		await OpenClientChannelAsync();

		await this.clientSession.CloseAsync(SshDisconnectReason.ByApplication);
		await this.serverSession.CloseAsync(SshDisconnectReason.ByApplication);

		await TaskExtensions.WaitUntil(() =>
			this.clientSession.Metrics.LatencyCurrentMs == 0 &&
			this.serverSession.Metrics.LatencyCurrentMs == 0).WithTimeout(Timeout);
	}

	[Fact]
	public async Task RecordSessionContour()
	{
		var clientContour = new SessionContour(256);
		var serverContour = new SessionContour(256);

		var clientContourTask = clientContour.CollectMetricsAsync(this.clientSession.Metrics);
		var serverContourTask = serverContour.CollectMetricsAsync(this.serverSession.Metrics);

		void ValidateContour(SessionContour contour)
		{
			// Normally the interval should be 1 second, but tests tests need to work
			// on very slow build machines where the interval could grow larger.
			Assert.Contains(
				contour.Interval,
				new[] { TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(4) });

			Assert.NotEqual(0, contour.IntervalCount);
			Assert.NotEqual(0, contour.LatencyMinMs.Sum());
			Assert.NotEqual(0, contour.LatencyMaxMs.Sum());
			Assert.NotEqual(0, contour.LatencyAverageMs.Sum());
			Assert.NotEqual(0, contour.BytesSent.Sum());
			Assert.NotEqual(0, contour.BytesReceived.Sum());

			for (int i = 0; i < contour.IntervalCount; i++)
			{
				Assert.True(contour.LatencyMinMs[i] <= contour.LatencyAverageMs[i]);
				Assert.True(contour.LatencyAverageMs[i] <= contour.LatencyMaxMs[i]);
			}
		}

		var channels = await OpenClientChannelAsync();

		await Task.Delay(TimeSpan.FromSeconds(1));
		byte[][] data = new[] { new byte[] { 1 } };
		await SendDataBetweenChannelsAsync(data, channels.Client, channels.Server);
		await SendDataBetweenChannelsAsync(data, channels.Server, channels.Client);
		await Task.Delay(TimeSpan.FromSeconds(1));
		await SendDataBetweenChannelsAsync(data, channels.Client, channels.Server);
		await SendDataBetweenChannelsAsync(data, channels.Server, channels.Client);
		await Task.Delay(TimeSpan.FromSeconds(1));
		await SendDataBetweenChannelsAsync(data, channels.Client, channels.Server);
		await SendDataBetweenChannelsAsync(data, channels.Server, channels.Client);

		await WaitForContourUpdateAsync(clientContour, clientContourTask);
		await WaitForContourUpdateAsync(serverContour, serverContourTask);

		ValidateContour(clientContour);
		ValidateContour(serverContour);
	}

	[Fact]
	public async Task ExpandContourIntervals()
	{
		var session = new SshClientSession(
			SshSessionConfiguration.Default, new TraceSource(nameof(MetricsTests)));
		var metrics = session.Metrics;

		var updateTask = this.sessionContour.CollectMetricsAsync(metrics);
		Assert.Equal(TimeSpan.FromSeconds(1), this.sessionContour.Interval);
		AddMessageReceived(2_000, 2);
		UpdateLatency(3_000, 16);
		UpdateLatency(3_500, 32);
		AddMessageSent(3_800, 1);
		AddMessageReceived(3_900, 3);
		await WaitForContourUpdateAsync(this.sessionContour, updateTask);
		Assert.Equal(TimeSpan.FromSeconds(1), this.sessionContour.Interval);
		Assert.Equal(new float[] { 0, 0, 0, 16 }, this.sessionContour.LatencyMinMs);
		Assert.Equal(new float[] { 0, 0, 0, 32 }, this.sessionContour.LatencyMaxMs);
		Assert.Equal(new float[] { 0, 0, 0, 24 }, this.sessionContour.LatencyAverageMs);
		Assert.Equal(new long[] { 0, 0, 0, 1 }, this.sessionContour.BytesSent);
		Assert.Equal(new long[] { 0, 0, 2, 3 }, this.sessionContour.BytesReceived);
		updateTask = this.sessionContour.CollectMetricsAsync(metrics);
		AddMessageSent(4_000, 1);
		UpdateLatency(4_500, 32);
		UpdateLatency(4_600, 16);
		await WaitForContourUpdateAsync(this.sessionContour, updateTask);
		Assert.Equal(TimeSpan.FromSeconds(2), this.sessionContour.Interval);
		Assert.Equal(new float[] { 0, 16, 16 }, this.sessionContour.LatencyMinMs);
		Assert.Equal(new float[] { 0, 32, 32 }, this.sessionContour.LatencyMaxMs);
		Assert.Equal(new float[] { 0, 24, 24 }, this.sessionContour.LatencyAverageMs);
		Assert.Equal(new long[] { 0, 1, 1 }, this.sessionContour.BytesSent);
		Assert.Equal(new long[] { 0, 5, 0 }, this.sessionContour.BytesReceived);
		updateTask = this.sessionContour.CollectMetricsAsync(metrics);
		AddMessageSent(8_000, 1);
		UpdateLatency(8_100, 32);
		AddMessageSent(12_000, 2);
		UpdateLatency(12_500, 64);
		await WaitForContourUpdateAsync(this.sessionContour, updateTask);
		Assert.Equal(TimeSpan.FromSeconds(4), this.sessionContour.Interval);
		Assert.Equal(new float[] { 16, 16, 32, 64 }, this.sessionContour.LatencyMinMs);
		Assert.Equal(new float[] { 32, 32, 32, 64 }, this.sessionContour.LatencyMaxMs);
		Assert.Equal(new float[] { 24, 24, 32, 64 }, this.sessionContour.LatencyAverageMs);
		Assert.Equal(new long[] { 1, 1, 1, 2 }, this.sessionContour.BytesSent);
		Assert.Equal(new long[] { 5, 0, 0, 0 }, this.sessionContour.BytesReceived);
		updateTask = this.sessionContour.CollectMetricsAsync(metrics);
		AddMessageSent(16_000, 10);
		await WaitForContourUpdateAsync(this.sessionContour, updateTask);
		Assert.Equal(TimeSpan.FromSeconds(8), this.sessionContour.Interval);
		Assert.Equal(new float[] { 16, 32, 0 }, this.sessionContour.LatencyMinMs);
		Assert.Equal(new float[] { 32, 64, 0 }, this.sessionContour.LatencyMaxMs);
		Assert.Equal(new float[] { 24, 48, 0 }, this.sessionContour.LatencyAverageMs);
		Assert.Equal(new long[] { 2, 3, 10 }, this.sessionContour.BytesSent);
		Assert.Equal(new long[] { 5, 0, 0 }, this.sessionContour.BytesReceived);
	}

	[Fact]
	public async Task ExportImportContour()
	{
		var session = new SshClientSession(
			SshSessionConfiguration.Default, new TraceSource(nameof(MetricsTests)));
		var metrics = session.Metrics;

		var updateTask = this.sessionContour.CollectMetricsAsync(metrics);
		AddMessageReceived(0, 2000);
		UpdateLatency(2_000, 16);
		UpdateLatency(3_000, 32);
		AddMessageSent(3_600, 1000);
		AddMessageReceived(3_800, 3000);
		AddMessageSent(4_000, 1);
		UpdateLatency(5_000, 32);
		UpdateLatency(5_200, 16);
		await WaitForContourUpdateAsync(this.sessionContour, updateTask);
		Assert.Equal(TimeSpan.FromSeconds(2), this.sessionContour.Interval);

		var result = this.sessionContour.Export();
		var resultBytes = Convert.FromBase64String(result);

		Assert.Equal(string.Join(" ", new byte[]
		{
				1,  // version
				5,  // metric count
				1,  // timeScale
				0,  // \
				0,  //  \
				0,  //   } value scales
				2,  //  /
				4,  // /
				1,  // \
				2,  //  \
				3,  //   } metric IDs
				11, //  /
				12, // /
				0,  // \
				0,  //  \
				0,  //   } interval 0
				0,  //  /
				125,// /
				16, // \
				32, //  \
				24, //   } interval 1
				250,//  /
				188,// /
				16, // \
				32, //  \
				24, //   } interval 2
				0,  //  /
				0,  // /
		}), string.Join(" ", resultBytes));

		var sessionContour2 = SessionContour.Import(result);
		Assert.Equal(3, sessionContour2.IntervalCount);
		Assert.Equal(TimeSpan.FromSeconds(2), sessionContour2.Interval);

		var result2 = sessionContour2.Export();
		var result2Bytes = Convert.FromBase64String(result2);
		Assert.Equal(string.Join(" ", resultBytes), string.Join(" ", result2Bytes));
	}

	private void AddMessageSent(long time, int size)
	{
		SessionContourOnMessageSentMethod.Invoke(
			this.sessionContour, new object[] { null, (time, size) });
	}

	private void AddMessageReceived(long time, int size)
	{
		SessionContourOnMessageReceivedMethod.Invoke(
			this.sessionContour, new object[] { null, (time, size) });
	}

	private void UpdateLatency(long time, float latency)
	{
		SessionContourOnLatencyUpdatedMethod.Invoke(
			this.sessionContour, new object[] { null, (time, latency) });
	}

	private async Task WaitForContourUpdateAsync(SessionContour sessionContour, Task t)
	{
		SessionContourSessionClosedMethod.Invoke(
			sessionContour, new object[] { null, EventArgs.Empty });
		await t.WithTimeout(Timeout);
	}

	private async Task<Channels> OpenClientChannelAsync()
	{
		await this.sessionPair.ConnectAsync().WithTimeout(Timeout);

		var serverChannelTask = this.serverSession.AcceptChannelAsync();
		var clientChannel = await this.clientSession.OpenChannelAsync().WithTimeout(Timeout);
		var serverChannel = await serverChannelTask.WithTimeout(Timeout);

		return new Channels
		{
			Client = clientChannel,
			Server = serverChannel,
		};
	}

	private static async Task SendDataBetweenChannelsAsync(
		byte[][] data,
		SshChannel channelA,
		SshChannel channelB)
	{
		MemoryStream receivingStream = null;
		TaskCompletionSource<byte[]> receivedCompletion = null;
		int expectedDataLength = 0;

		EventHandler<Buffer> dataReceivedHandler = (sender, bytes) =>
		{
			receivingStream.Write(bytes.Array, bytes.Offset, bytes.Count);
			if (receivingStream.Length >= expectedDataLength)
			{
				receivedCompletion.SetResult(receivingStream.ToArray());
			}

			channelB.AdjustWindow((uint)bytes.Count);
		};
		channelB.DataReceived += dataReceivedHandler;

		for (int i = 0; i < data.Length; i++)
		{
			receivingStream = new MemoryStream();
			receivedCompletion = new TaskCompletionSource<byte[]>();
			expectedDataLength = data[i].Length;

			await channelA.SendAsync(data[i], CancellationToken.None);

			var receivedData = await receivedCompletion.Task.WithTimeout(
				TimeSpan.FromMilliseconds(4 * Timeout.TotalMilliseconds));
			Assert.Equal(data[i], receivedData);
		}

		channelB.DataReceived -= dataReceivedHandler;
	}

	[DebuggerStepThrough]
	private class Channels
	{
		public SshChannel Client { get; set; }
		public SshChannel Server { get; set; }
	}
}
